flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayOfStringArgumentTypeStrategy.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.types.inference.strategies;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.functions.FunctionDefinition;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.inference.ArgumentTypeStrategy;
+import org.apache.flink.table.types.inference.CallContext;
+import org.apache.flink.table.types.inference.Signature.Argument;
+import org.apache.flink.table.types.logical.ArrayType;
+import org.apache.flink.table.types.logical.LogicalTypeRoot;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+/** Strategy for an argument that must be an array of strings. */
+@Internal
+public final class ArrayOfStringArgumentTypeStrategy implements ArgumentTypeStrategy {
+
+    private final String constraintMessage;
+
+    public ArrayOfStringArgumentTypeStrategy(String constraintMessage) {
+        this.constraintMessage = constraintMessage;
+    }
+
+    @Override
+    public Optional<DataType> inferArgumentType(
+            CallContext callContext, int argumentPos, boolean throwOnFailure) {
+
+        final List<DataType> actualDataTypes = callContext.getArgumentDataTypes();
+        DataType actualType = actualDataTypes.get(argumentPos);
+
+        // Check if the type is an array and its element is string
+        if (actualType.getLogicalType().getTypeRoot() == LogicalTypeRoot.ARRAY
+                && (((ArrayType) actualType.getLogicalType()).getElementType().getTypeRoot()
+                                == LogicalTypeRoot.VARCHAR
+                        || ((ArrayType) actualType.getLogicalType()).getElementType().getTypeRoot()
+                                == LogicalTypeRoot.CHAR)) {
+            return Optional.of(actualType);
+        }
```

Comment:
We should rather check if we can implicitly cast to `VARCHAR`.  You can reuse: https://github.com/apache/flink/blob/fbc086f2fd63dac4af3e104ba4011e864781139a/flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/StrategyUtils.java#L56

Reply:
fixed it